### PR TITLE
Utility to check whether a function is compiled with framepointers

### PR DESCRIPTION
**Pull request by florian-kuebler**
The checker checks a functions Prologue and is also looking for a matching Epilogue.
Leaf functions will be no target to that check, i.e. it is fine to have leaf functions without a correct Prologue/Epilogue.

### CODE REVIEW

File: OrbitCore/CMakeLists.txt [hunk 1]
```diff
@@ -5,7 +5,7 @@
 cmake_minimum_required(VERSION 3.15)
 
 project(OrbitCore)
-add_library(OrbitCore STATIC)
+add_library(OrbitCore STATIC FunctionFramepointerValidator.cpp FunctionFramepointerValidator.h)
```
Comment (dpallotti) [inline OrbitCore/CMakeLists.txt:8]: These don't belong here.

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  instructions_ = nullptr;
+  instructions_count_ =
+      cs_disasm(handle, code, code_size, 0, 0, &instructions_);
+}
+
+FunctionFramepointerValidator::~FunctionFramepointerValidator() {
+  cs_free(instructions_, instructions_count_);
+}
+
+bool FunctionFramepointerValidator::isCallInstruction(
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.cpp:22]: Method names should be PascalCase, not camelCase.

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+bool FunctionFramepointerValidator::isBasePointer(uint16_t reg) {
+  return reg == X86_REG_BP || reg == X86_REG_EBP || reg == X86_REG_RBP;
+}
+
+bool FunctionFramepointerValidator::isStackPointer(uint16_t reg) {
+  return reg == X86_REG_SP || reg == X86_REG_ESP || reg == X86_REG_RSP;
+}
+
+// we will actually only check, whether there is a correct epilogue
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.cpp:43]: Can you rephrase better? Does it mean that the existence of one epilogue is enough?

File: OrbitCore/FunctionFramepointerValidator.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+  // FunctionFramepointerValidator is neither copyable nor movable.
+  FunctionFramepointerValidator(const FunctionFramepointerValidator&) = delete;
+  FunctionFramepointerValidator& operator=(
+      const FunctionFramepointerValidator&) = delete;
+
+  bool validate();
+
+ private:
+  bool isCallInstruction(const cs_insn& instruction);
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.h:31]: Most of these methods can be static.

File: OrbitCore/FunctionFramepointerValidatorTest.cpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
+#include "FunctionFramepointerValidator.h"
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidatorTest.cpp:2]: The copyright should be at the very top.

File: OrbitCore/FunctionFramepointerValidatorTest.cpp [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <capstone/capstone.h>
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
+#define WITH_FP "\x55\x89\xE5\x83\xC0\x01\xE8\x77\x00\x00\x00\x89\xEC\x5D"
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidatorTest.cpp:10]: Do these need to be `define`s? Can they just be constant?

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+bool FunctionFramepointerValidator::validateEpilogue() {
+  bool found_epilogue = false;
+  for (size_t i = 0; i < instructions_count_ - 1; i++) {
+    if (isMovInstruction(instructions_[i]) &&
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidator.cpp:47]: do you check if there is ret after mov and pop?

File: OrbitCore/FunctionFramepointerValidator.h [hunk 8]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_CORE_FUNCTION_FRAMEPOINTER_VALIDATOR_H_
+#define ORBIT_CORE_FUNCTION_FRAMEPOINTER_VALIDATOR_H_
+
+#include <capstone/capstone.h>
+
+/*
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.h:10]: We are not really using this style for class comments. Use `//` on every line.

File: OrbitCore/FunctionFramepointerValidator.h [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+ * The validator checks the functions Prologue and also whether an Epilogue
+ * exists.
+ */
+class FunctionFramepointerValidator {
+ public:
+  FunctionFramepointerValidator(csh handle, const uint8_t* code,
+                                size_t code_size);
+  virtual ~FunctionFramepointerValidator();
+
+  // FunctionFramepointerValidator is neither copyable nor movable.
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.h:23]: Then delete the move constructor and assignment operator as well.

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+bool FunctionFramepointerValidator::isLeafFunction() {
+  for (size_t i = 0; i < instructions_count_; i++) {
+    if (isCallInstruction(instructions_[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool FunctionFramepointerValidator::validate() {
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidator.cpp:113]: how about exception for _start which does some special things?

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+  bool found_epilogue = false;
+  for (size_t i = 0; i < instructions_count_ - 1; i++) {
+    if (isMovInstruction(instructions_[i]) &&
+        instructions_[i + 1].id == X86_INS_POP) {
+      cs_regs regs_read_first, regs_write_first, regs_read_second,
+          regs_write_second;
+      uint8_t read_count_first, write_count_first, read_count_second,
+          write_count_second;
+      // check for "mov esp, ebp; pop ebp"
+      if (  // first instruction
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.cpp:54]: This condition is unreadable. I would split it in a few boolean variables, *at least* one for "first instruction" and one for "second instruction", but possibly more.

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 10 above)
Comment (dimitry-) [reply OrbitCore/FunctionFramepointerValidator.cpp:113]: Actually never mind, this method probably is not called at all for _start

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        !isBasePointer(regs_write[0])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool FunctionFramepointerValidator::validateFramePointers() {
+  return validateEpilogue() && validatePrologue();
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.cpp:101]: Isn't it weird to have `validatePrologue` second?

Comment (dpallotti) [reply OrbitCore/FunctionFramepointerValidator.cpp:101]: Same goes for the order of declaration and definition in the class.

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+  for (size_t i = 0; i < instructions_count_; i++) {
+    if (isCallInstruction(instructions_[i])) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool FunctionFramepointerValidator::validate() {
+  if (instructions_count_ > 0) {
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.cpp:114]: What about an early exit?
```
if (instruction_count_ == 0) {
  ...
  return false;
}

<unindented code>
```

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 3 above)
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.cpp:43]: Pick up the habit of writing comments with a starting capital letter and a final period (per the style guide) before it's too late.

File: OrbitCore/FunctionFramepointerValidator.h (hunk 9 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.h:21]: Okay, Done.

File: OrbitCore/FunctionFramepointerValidatorTest.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidatorTest.cpp]: Sure, thanks, I adjusted them accordingly. The defines were basically from the developer guide of capstone.

Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidatorTest.cpp]: Thanks for spotting this. Please see above.

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 10 above)
Comment (dimitry-) [reply OrbitCore/FunctionFramepointerValidator.cpp]: yes. And also we probably do not even care what it is called because the way to add expection would be to examine entry_point in the Elf_Ehdr

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 14]
```diff
@@ -19,113 +19,119 @@ FunctionFramepointerValidator::~FunctionFramepointerValidator() {
...
   return true;
 }
 
-bool FunctionFramepointerValidator::validateFramePointers() {
-  return validateEpilogue() && validatePrologue();
+// We will actually only check, whether there is any correct epilogue.
+// It might be the case, that there are multiple function returns and that not
+// all are correct. However, we would not expect a compiler to produce this,
+// and for hand written assembly, we accept wrong unwinding results.
+bool FunctionFramepointerValidator::ValidateEpilogue() {
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidator.cpp:73]: to make it more readable use early returns..

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 15]
```diff
@@ -19,113 +19,119 @@ FunctionFramepointerValidator::~FunctionFramepointerValidator() {
...
+  bool found_epilogue = false;
+  for (size_t i = 0; i < instructions_count_ - 1; i++) {
+    if (IsMovInstruction(instructions_[i]) &&
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidator.cpp:76]: here you can do
if (!IsMov ... ) {
  continue;
}

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 16]
```diff
@@ -19,113 +19,119 @@ FunctionFramepointerValidator::~FunctionFramepointerValidator() {
...
+          regs_write_second;
+      uint8_t read_count_first, write_count_first, read_count_second,
+          write_count_second;
+
+      // Check first instruction to be "mov esp, ebp"
+      bool succeeded_regs_access_first =
+          cs_regs_access(handle_, &instructions_[i], regs_read_first,
+                         &read_count_first, regs_write_first,
+                         &write_count_first) == 0;
+
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidator.cpp:88]: if succeeded_regs_access_first is false you can probably continue... since in this case checks below do not make sense, right?

and so on..

File: OrbitCore/FunctionFramepointerValidatorTest.cpp [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+// found in the LICENSE file.
+#include "FunctionFramepointerValidator.h"
+
+#include <capstone/capstone.h>
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+
+#define NO_FP \
+  "\x29\x25\x00\x00\x00\x00\xE8\xFD\xFF\xFF\xFF\x01\x25\x00\x00\x00\x00\xC3"
+#define LEAF "\x29\x24\x25\x00\x00\x00\x00\x48\x01\x24\x25\x00\x00\x00\x00\xC3"
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidatorTest.cpp:12]: forgot to remove these?

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 14 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp:73]: Okay, sure.

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 15 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp]: Good point

File: OrbitCore/FunctionFramepointerValidator.h [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+  FunctionFramepointerValidator(const FunctionFramepointerValidator&) = delete;
+  FunctionFramepointerValidator& operator=(
+      const FunctionFramepointerValidator&) = delete;
+  FunctionFramepointerValidator(FunctionFramepointerValidator&&) = delete;
+  FunctionFramepointerValidator& operator=(
+      const FunctionFramepointerValidator&&) = delete;
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.h:27]: Remove this const as it's an rvalue-ref.

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 16 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp]: Thanks for spotting this. I made them if(...) continues.

File: OrbitCore/FunctionFramepointerValidator.h (hunk 4 above)
Comment (dpallotti) [reply OrbitCore/FunctionFramepointerValidator.h]: You mean free-standing functions? I would say it's up to you. They should just not be instance methods.

File: OrbitCore/FunctionFramepointerValidatorTest.cpp (hunk 17 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidatorTest.cpp]: Got me. Thanks

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// We will actually only check, whether there is any correct epilogue.
+// It might be the case, that there are multiple function returns and that not
+// all are correct. However, we would not expect a compiler to produce this,
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.cpp:71]: I'm pretty sure compilers produce multiple ret instructions in optimized code (maybe not together with "mov esp, ebp" and "pop ebp", but for the sake of this comment...)

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+                         &read_count_second, regs_write_second,
+                         &write_count_second) == 0;
+
+      bool correct_regs_access_second =
+          succeeded_regs_access_second && read_count_second == 1 &&
+          IsStackPointer(regs_read_second[0]) && write_count_second == 2 &&
+          IsStackPointer(regs_write_second[0]) &&
+          IsBasePointer(regs_write_second[1]);
+
+      // Are both instructions as expected? If so, we found an epilogue
```
Comment (dpallotti) [inline OrbitCore/FunctionFramepointerValidator.cpp:106]: What about the "leave" instruction, which is equivalent to "mov esp, ebp","pop ebp"?

File: OrbitCore/FunctionFramepointerValidator.h (hunk 18 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.h]: Okay. Thanks.

File: OrbitCore/FunctionFramepointerValidator.h (hunk 4 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.h]: We should discuss this offline at some later point in time. For now, I will just make them offline.

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 20 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp]: Good point. I wasn't aware of this instruction. I added an early return if we find this instruction.

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+      continue;
+    }
+    if (instructions_[i + 1].id != X86_INS_POP) {
+      continue;
+    }
+
+    cs_regs regs_read_first, regs_write_first, regs_read_second,
+        regs_write_second;
+    uint8_t read_count_first, write_count_first, read_count_second,
+        write_count_second;
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidator.cpp:90]: naming here is still a bit confusing.

maybe remove first and second and reuse them.
regs_read = source_registers
regs_write = destination registers

Comment (dimitry-) [reply OrbitCore/FunctionFramepointerValidator.cpp:90]: On other hand for pop instruction it is not even that... maybe use different naming there

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 7 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp]: In capstone there is: 
```// all return instructions
X86_GRP_RET```

So it should be safe to use this group, as done for calls.

But I don't get you comment with pop rip. Could you elaborate?

Comment (dimitry-) [reply OrbitCore/FunctionFramepointerValidator.cpp]: if you have return address on the top of the stack pop rip is effectively a ret. The question is if amd64 lets regular instructions modify rip (or directly assign something to rip), for example arm32 lets you do that.

From what I understand amd64 does not allow that so we should be fine with checking RET

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 21 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp:90]: I would be fine with just having `regs_read` and `regs_write`

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 7 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp]: Ah thanks for clarifying. Well lets implement the check for RET, we would only produce false positive warnings if pop rip would be possible though.

**Comment by florian-kuebler** [comment]
I updated the PR with handling of the RET instructions as well as simplification of the naming.
Further, I added handling of ENTER/LEAVE instructions.

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 22]
```diff
@@ -47,6 +55,10 @@ bool FunctionFramepointerValidator::ValidatePrologue() {
   // check the first instruction: it must be "push ebp"
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidator.cpp:55]: or enter apparently...

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 23]
```diff
@@ -47,6 +55,10 @@ bool FunctionFramepointerValidator::ValidatePrologue() {
...
   if (cs_regs_access(handle_, &instructions_[0], regs_read, &read_count,
                      regs_write, &write_count) == 0) {
+    if (instructions_[0].id == X86_INS_ENTER) {
+      return true;
+    }
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidator.cpp:60]: ok, this probably introduces another state... the question here is, is it valid to use ENTER without LEAVE and LEAVE without ENTER.

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 24]
```diff
@@ -71,9 +83,10 @@ bool FunctionFramepointerValidator::ValidatePrologue() {
 // all are correct. However, we would not expect a compiler to produce this,
 // and for hand written assembly, we accept wrong unwinding results.
 bool FunctionFramepointerValidator::ValidateEpilogue() {
-  for (size_t i = 0; i < instructions_count_ - 1; i++) {
+  for (size_t i = 0; i < instructions_count_ - 2; i++) {
     // "leave" is equivalent to "mov esp, ebp" "pop ebp"
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidator.cpp:87]: not always - apparently it depends on nesting attribute. it could do more if I understand correctly

File: OrbitCore/FunctionFramepointerValidatorTest.cpp [hunk 25]
```diff
@@ -11,7 +11,7 @@ namespace {
 
 constexpr uint8_t kFunctionWithFP[] = {0x55, 0x89, 0xE5, 0x83, 0xC0,
                                        0x01, 0xE8, 0x77, 0x00, 0x00,
-                                       0x00, 0x89, 0xEC, 0x5D};
+                                       0x00, 0x89, 0xEC, 0x5D, 0xC3};
 constexpr uint8_t kFunctionWithoutFP[] = {0x29, 0x25, 0x00, 0x00, 0x00, 0x00,
                                           0xE8, 0xFD, 0xFF, 0xFF, 0xFF, 0x01,
                                           0x25, 0x00, 0x00, 0x00, 0x00, 0xC3};
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidatorTest.cpp:17]: do you want to add test for ENTER and LEAVE - including invalid ones maybe, depending on what we decide.

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 23 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp:62]: As discussed offline, this would be strange hand-crafted assembly, which would silently accept here.

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 22 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp]: Thanks.

File: OrbitCore/FunctionFramepointerValidator.cpp (hunk 24 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp:94]: As discussed offline, we would still accept this here.

File: OrbitCore/FunctionFramepointerValidatorTest.cpp (hunk 25 above)
Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidatorTest.cpp:17]: Added a test for this and also for tail calls. Thanks.

File: OrbitCore/FunctionFramepointerValidator.cpp [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+// It might be the case, that there are multiple function returns and that not
+// all are correct. However, we would not expect a compiler to produce this,
+// and for hand written assembly, we accept wrong unwinding results.
+//
+// When functions are tail call optimized, the callee might not have a "ret"
+// after the epilogue. In this case we just assume that a "jump" after the
+// epilogue is the return to the caller.
+// TODO(kuebler): Better handling for tail call optimization
+bool FunctionFramepointerValidator::ValidateEpilogue() {
+  for (size_t i = 0; i < instructions_count_ - 2; i++) {
```
Comment (dimitry-) [inline OrbitCore/FunctionFramepointerValidator.cpp:93]: - 2 here could be invalid in the case when last 2 instructions are LEAVE and RET, right? Maybe change it to while true and introduce instruction iterator to avoid such corner cases here.

Comment (florian-kuebler) [reply OrbitCore/FunctionFramepointerValidator.cpp]: Thanks for spotting. I fixed this accordingly.
